test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller.go

```diff
@@ -230,25 +279,178 @@ func (r *DockerMachinePoolReconciler) reconcileNormal(ctx context.Context, clust
+	dockerMachineList, err := getDockerMachines(ctx, r.Client, *cluster, *machinePool, *dockerMachinePool)
+	if err != nil {
+		return err
+	}
+
+	log.V(2).Info("DockerMachineList kind is", "kind", dockerMachineList.GetObjectKind())
+
+	instanceNameSet := map[string]struct{}{}
+	for _, instance := range instances {
+		instanceNameSet[instance.InstanceName] = struct{}{}
+	}
+
+	for i := range dockerMachineList.Items {
+		dockerMachine := &dockerMachineList.Items[i]
+		if _, ok := instanceNameSet[dockerMachine.Spec.InstanceName]; !ok {
+			machine, err := util.GetOwnerMachine(ctx, r.Client, dockerMachine.ObjectMeta)
+			if err != nil {
+				return errors.Wrapf(err, "failed to get owner Machine for DockerMachine %s/%s", dockerMachine.Namespace, dockerMachine.Name)
+			}
+			if machine == nil {
+				return errors.Errorf("DockerMachine %s/%s has no parent Machine, will reattempt deletion once parent Machine is present", dockerMachine.Namespace, dockerMachine.Name)
+			}
+
+			if err := r.Client.Delete(ctx, machine); err != nil {
+				return errors.Wrapf(err, "failed to delete orphaned DockerMachine %s/%s", dockerMachine.Namespace, dockerMachine.Name)
+			}
+		} else {
+			log.V(2).Info("Keeping DockerMachine, nothing to do", "dockerMachine", dockerMachine.Name, "namespace", dockerMachine.Namespace)
+		}
+	}
+
+	return nil
+}
+
+// CreateDockerMachinesIfNotExists creates a DockerMachine for each instance returned by the node pool if it doesn't exist.
+func (r *DockerMachinePoolReconciler) CreateDockerMachinesIfNotExists(ctx context.Context, cluster *clusterv1.Cluster, machinePool *expv1.MachinePool, dockerMachinePool *infraexpv1.DockerMachinePool, instances []docker.NodePoolInstance) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.V(2).Info("Creating missing machines", "dockerMachinePool", dockerMachinePool.Name, "namespace", dockerMachinePool.Namespace, "instances", instances)
+
+	dockerMachineList, err := getDockerMachines(ctx, r.Client, *cluster, *machinePool, *dockerMachinePool)
+	if err != nil {
+		return err
+	}
+
+	instanceNameToDockerMachine := make(map[string]infrav1.DockerMachine)
+	for _, dockerMachine := range dockerMachineList.Items {
+		instanceNameToDockerMachine[dockerMachine.Spec.InstanceName] = dockerMachine
+	}
+
+	for _, instance := range instances {
+		if _, exists := instanceNameToDockerMachine[instance.InstanceName]; exists {
+			continue
+		}
+
+		labels := map[string]string{
+			clusterv1.ClusterNameLabel:     cluster.Name,
+			clusterv1.MachinePoolNameLabel: machinePool.Name,
+		}
+		labels[clusterv1.MachinePoolNameLabel] = machinePool.Name
+		dockerMachine := &infrav1.DockerMachine{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:    dockerMachinePool.Namespace,
+				GenerateName: fmt.Sprintf("%s-", dockerMachinePool.Name),
+				Labels:       labels,
+				Annotations:  make(map[string]string),
+				// Note: This DockerMachine will be owned by the DockerMachinePool until the MachinePool controller creates its parent Machine.
```

Comment:
This is implemented as the MachinePool controller automatically sets the ownerRef. I just left the note here so people can expect that.

Reply:
I think we still need to set the parent to infra machinepool or machinepool here as otherwise there are chances of orphaned objects right? We saw orphaned objects in our testing, as there is a period of time between we create the machine pool infra machine and we se the owner machine taht something can happen and the infra machines can remain orphaned  i.e. without parent and hence wont be deleted.

Reply:
Gotcha. I was under the impression that because the DockerMachines were created in the DockerMachinePool controller, they would be owned by the DockerMachinePool automatically. But in any case I suppose it can't hurt to set the ownerRef to be safe.